critwall/src/main/java/org/opencadc/critwall/FileSyncJob.java

```diff
@@ -306,105 +301,147 @@ private void doSync() {
+            URL transferURL = regClient.getServiceURL(resource, Standards.SI_LOCATE, am);
+            if (transferURL == null) {
+                transferURL = regClient.getServiceURL(resource, Standards.VOSPACE_SYNC_21, am);
             }
+            log.debug("certURL: " + transferURL);
+    
+            // request all protocols that can be used
+            List<Protocol> protocolList = new ArrayList<>();
+            protocolList.add(new Protocol(VOS.PROTOCOL_HTTPS_GET));
+            protocolList.add(new Protocol(VOS.PROTOCOL_HTTP_GET));
+            if (!AuthMethod.ANON.equals(am)) {
+                Protocol httpsAuth = new Protocol(VOS.PROTOCOL_HTTPS_GET);
+                httpsAuth.setSecurityMethod(Standards.getSecurityMethod(am));
+                protocolList.add(httpsAuth);
+            }
+    
+            Transfer transfer = new Transfer(artifact, Direction.pullFromVoSpace, protocolList);
+            transfer.version = VOS.VOSPACE_21;
+    
+            TransferWriter writer = new TransferWriter();
+            ByteArrayOutputStream out = new ByteArrayOutputStream();
+            writer.write(transfer, out);
+            FileContent content = new FileContent(out.toByteArray(), "text/xml");
+            log.warn("xml file content to be posted: " + transfer);
+    
+            log.debug("artifact path: " + artifact.getPath());
+            HttpPost post = new HttpPost(transferURL, content, true);
+            post.prepare();
+            log.debug("post prepare done");
+    
+            TransferReader reader = new TransferReader();
+            Transfer t = reader.read(post.getInputStream(), null);
+            return t.getProtocols();
         }
-
-        return urlList;
     }
 
-    private StorageMetadata syncArtifact(Artifact a, List<URL> urls)
-        throws ByteLimitExceededException, StorageEngageException, InterruptedException, WriteException, IllegalArgumentException {
-
-        StorageMetadata storageMeta = null;
-        Iterator<URL> urlIterator = urls.iterator();
-
+    private StorageMetadata syncArtifact(Artifact a, List<Protocol> protocols)
+        throws ByteLimitExceededException, StorageEngageException, InterruptedException, 
+            WriteException, IllegalArgumentException, PrivilegedActionException, MalformedURLException {
+        Iterator<Protocol> urlIterator = protocols.iterator();
         while (urlIterator.hasNext()) {
-            URL u = urlIterator.next();
+            Protocol p = urlIterator.next();
+            URL u = new URL(p.getEndpoint());
             log.debug("trying " + u);
+            Subject subj = subject;
 
+            if (p.getSecurityMethod() == null || p.getSecurityMethod().equals(Standards.getSecurityMethod(AuthMethod.ANON))) {
+                subj = AuthenticationUtil.getAnonSubject();
+            }
+
+            try {
+                return Subject.doAs(subj, new SyncArtifact(u, a));
+            } catch (RuntimeException ex) {
```

Comment:
Using the message in a RuntimeException for application logic is really bad practice. 

The previous code used the different catch blocks to remove a url (now protocol) from the iterator and it was clear which exceptions were judged transient (don't remove) vs fatal for that endpoint (remove) vs fatal for the job (don't catch or rethrow). The SyncArtifact.run() declares a bunch of exceptions, so just throw them and let the caller sort it out as before.

`Subject.doAs(...)` does make it harder to do error handling but this is now worse than it was before: fragile and easily broken by text changes. It's possible that you moved too much of the code from the iterator loop into the SyncArtifact action: the chosen subject is only needed to call `get.prepare()` and after that A&A has succeeded and the InputStream can be read. that means a lot of error handling for back end failures has to also transition out of that Subject.doAs(...) ... 

Please re-think the error handling here. Ask (slack) if you want advice on how to reorganise this.

Reply:
Refactored SyncArtifact to handle all Exceptions in the caller. Currently the Java compiler complains about catching a couple of Exceptions in the caller that are not thrown in SyncArtifact. However those Exceptions are being thrown in SyncArtifact.run().